### DISCUSSION TITLE
vscode addon for formatting and highlighting

### DISCUSSION THREAD

**Discussion by Natejoestev** [moudey/Shell#81] [Ideas]
it would be nice if you made a vscode addon for Nilesoft Shell Script (`.shl`).
it would add simple text highlighting and support for:
- collapsing `{...}` and multi-line items
- maybe IntelliSense
<br>

(also, it would be nice if you had a discord server👍)

**Comment by Natejoestev** [comment]
i'm looking into it and i think i might beable to do it myself.

**Comment by moudey** [comment]
This would be a really great contribution

**Comment by ChristopheL77** [comment]
Hi, I've made a vscode extension for that. With a grammar, if you wanna try.
I'm new to this, but it seems to work well :-)
![Screenshot 2023-05-25 142811](https://github.com/moudey/Shell/assets/16576956/f4646cd9-6af0-4d87-bdcf-1118fc218584)

Folding works, comments and hightlighting of some elements is supported.

I can give you the code if you want to use it as a starter.

[vsix extension file](https://transfer.pcloud.com/download.html?code=5Ztoi6VZAYdFMQflG64ZSWbNZIkkYOYuEUJbkwAilI3gFdJklFLvX)

**Reply by moudey** [reply]
This is a really great contribution

**Reply by moudey** [reply]
You can add this extension to the vscode folder inside the extensions folder

**Reply by ChristopheL77** [reply]
I don't have rights to create a branch.

**Comment by Natejoestev** [comment]
nice, i will use that.

**Comment by ChristopheL77** [comment]
Fixed variables not colored:
[updated vsix extension](https://transfer.pcloud.com/download.html?code=5ZI3r6VZAYdFMQflG64ZSWbNZWxw20Ne07KmUghCrUTNKAfq0QS9X)

Also added snippets (me/it) for menu/item

**Reply by ChristopheL77** [reply]
[Updated vsix extension](https://transfer.pcloud.com/download.html?code=5ZlrXEVZAYdFMQflG64ZSWbNZWEW84g8AYM0RpbKICawz9zErDhLk)

**Reply by ChristopheL77** [reply]
New update with all set sections from the documentation

[Vsix 0.1.0](https://transfer.pcloud.com/download.html?code=5ZDtbEVZAYdFMQflG64ZSWbNZ4RYFbBJkYQ7eNzme22lsmbhuMm17)

**Reply by Natejoestev** [reply]
i'm sorry, but it's not doing anything. :/

**Reply by Natejoestev** [reply]
@ChristopheL77 the newest versions `0.0.5` and `0.1.0` don't do anything.
I can't get the old ones, link expired

**Reply by Natejoestev** [reply]
i understand what you did.
the file extension is **not** `.nss` it's `.shl`

**Reply by Natejoestev** [reply]
line 22, change to `.shl` in package.json

**Reply by ChristopheL77** [reply]
You mean the extensions of files for nilesoft shell ?
I think it is .nss because it is what is specified in documentation and also in the example files delivered by the setup : [Doc](https://nilesoft.org/) 
By the way, you can switch to nilesoft in the vscode state bar even if it is the wrong extension no ?

**Reply by ChristopheL77** [reply]
> @ChristopheL77 the newest versions `0.0.5` and `0.1.0` don't do anything. I can't get the old ones, link expired

Sorry for expired links. I thought I could put it in github but I can't create a branch right now.

**Reply by ChristopheL77** [reply]
I published it on the market place, you should be able to install it from vscode now 

![image](https://github.com/moudey/Shell/assets/16576956/f4d552e4-d7ed-4e35-b9a5-408c648d6c87)

**Comment by Natejoestev** [comment]
that is weird, my nilesoft download has `shell.shl` as the file.
also, i found a bug when you have smth like `a='@myFnc("b")'` it doesn't like the quotes or smth.

**Reply by ChristopheL77** [reply]
Is it an old version ?

**Reply by ChristopheL77** [reply]
>i found a bug when you have smth like `a='@myFnc("b")'` it doesn't like the quotes or smth.

Yes I didn't find a regular expression to match these expressions because nss language allows to break item/menu properties into multiple lines mixed with properties in same line and there is not an end of statement char as far as I know. I'm still trying though ;)
I mean we can do
```
menu(prop1=abc())
menu(prop1=abc()
prop2=def
)
menu(prop1=abc(aa()) prop2=def
)
menu(prop1=abc prop2=def()
)
```
And I have to identity the last parenthesis to know where the menu properties ends.

Here, the right parenthesis may be interpreted as the end of menu or item properties that's why it doesn't highlight well after.

**Reply by Natejoestev** [reply]
i have 1.7 rn. how do i update it?

**Reply by ChristopheL77** [reply]
winget update nilesoft.shell

Or others methods : https://nilesoft.org/download
Backup your config files before if you updated them.

**Reply by Natejoestev** [reply]
could you add the nilesoft icon to the file extension?

**Reply by ChristopheL77** [reply]
>  i found a bug when you have smth like `a='@myFnc("b")'` it doesn't like the quotes or smth.

I think it is fixed now in v1.1.0
I learned about regular expression recursion ;-)

**Comment by ChristopheL77** [comment]
Yes if @moudey allows me to use his icon.

**Reply by moudey** [reply]
Yes you can

**Reply by ChristopheL77** [reply]
Cool ;). Updated in v1.0.1

**Reply by Natejoestev** [reply]
that is not what i meant. i meant for the files so you can see that they are nilesoft shell files instead of just this:
![image](https://github.com/moudey/Shell/assets/85044853/78d3a2b2-b4b8-4ccd-aeca-aa563eb8f4d5)

**Reply by ChristopheL77** [reply]
Done in 1.2.0 ;)

**Comment by ChristopheL77** [comment]
Added basic completion in 1.3.0

![image](https://github.com/moudey/Shell/assets/16576956/f62cd5e4-961a-4344-943b-21784d5422d9)

![image](https://github.com/moudey/Shell/assets/16576956/711b7f25-0b52-48df-865d-f3c97c817efe)

**Reply by ChristopheL77** [reply]
Updated to 1.5.0 with autocompletion with documentation (based on nilesoft doc)

**Reply by Natejoestev** [reply]
@ChristopheL77 i would like to get access to the extension repo so i can fix some things.

**Reply by ChristopheL77** [reply]
What do you want to fix ?
Send me your email address so I can add you to project.

**Reply by Natejoestev** [reply]
are you in the nilesoft discord?
also, i don't think you need an email, you can enter `username, full name, or email`.
just use my username @Natejoestev

**Reply by Natejoestev** [reply]
unless your planning to add me as a contributor to the marketplace.

**Comment by ChristopheL77** [comment]
Repo is not in github

Le lun. 11 déc. 2023, 02:20, Natejoestev ***@***.***> a
écrit :

**Reply by Natejoestev** [reply]
can we talk on discord?

**Reply by Natejoestev** [reply]
my username is `@natejoestev`

**Comment by JoyHak** [comment]
Check out my Notepad++ theme for Shell & Python: https://discord.com/channels/1106387012707168318/1241063722449899712

**Comment by RubicBG** [comment]
@ChristopheL77 do you plan make the NS vscode extension even better?

**Comment by ChristopheL77** [comment]
Hi,

Not really.

What do you need ?

**Comment by RubicBG** [comment]
Тhanks to text highlighting the code is easily recognizable - without it it would be much more difficult - very good job. But I personally use regex, reg paths, ps commands ... - sometimes they make highlighting "stop" for a few lines, or change the coloring, some new commands are not colored, auto-completion only works on basic commands ... definitely many things can be improved. If you decide to work on improving the extension I would like to help

**Comment by ChristopheL77** [comment]
Thanks.

I know it is perfectible.  It is textmate based. I did that as a hobby.
To go further, I think we would need a language server but I don't have
time for this.
I can give you access to source code, it is hosted on azure devops (I need an email address).


Christophe

**Reply by RubicBG** [reply]
I won't have time in the next few months. I'll need time to get familiar with azure devops and the extension structure before I start changing anything on my own. You can add me to the project now or I'll write to you again when I have time to take on the project more seriously. Find me in discord with the same name (I don't want to use the program page as a general chat)

**Comment by pa-0** [comment]
This is really awesome.  How difficult would it be to release this as a syntax definition package for Sublime Text?

**Reply by ChristopheL77** [reply]
I don't know at all. I never used sublime nor did an extension for it. Sorry.

**Reply by pa-0** [reply]
Is your extension open source? If so I'd be happy to fork and port it over to sublime text for those of us who don't use vscode

**Reply by ChristopheL77** [reply]
I can give you access to source code, it is hosted on azure devops (I need an email address).

**Reply by pa-0** [reply]
Sure.  Here is a throwaway email I use (to avoid publishing my primary): inthat650@gmail.com

**Reply by pa-0** [reply]
I will maintain one branch for editor-agnostic commits in case you'd like to merge any of them into your repo.
